File: frontend/src/lib/types/ws-messages.ts
/**
 * WebSocket message type definitions.
 *
 * These types define the wire contract between frontend and backend.
 * Both sides MUST stay in sync — the Python counterpart is at
 * backend/src/openroast/models/ws_messages.py.
 */

// ──────────────────────────────────────────────
// Shared enums
// ──────────────────────────────────────────────

export type RoastEventType =
  | "CHARGE"
  | "DRY_END"
  | "FCs"
  | "FCe"
  | "SCs"
  | "SCe"
  | "DROP"
  | "COOL"
  | "TP";

export type SessionState = "idle" | "monitoring" | "recording" | "finished";

export type AlarmSeverity = "info" | "warning" | "critical";

export type DriverState = "disconnected" | "connecting" | "connected" | "error";

// ──────────────────────────────────────────────
// Server → Client messages
// ──────────────────────────────────────────────

export interface TemperatureMessage {
  type: "temperature";
  timestamp_ms: number;
  et: number;
  bt: number;
  et_ror: number;
  bt_ror: number;
  extra_channels: Record<string, number>;
}

export interface EventMessage {
  type: "event";
  event_type: RoastEventType;
  timestamp_ms: number;
  auto_detected: boolean;
  bt_at_event: number;
  et_at_event: number;
}

export interface StateMessage {
  type: "state";
  state: SessionState;
  previous_state: SessionState;
}

export interface AlarmMessage {
  type: "alarm";
  alarm_id: string;
  message: string;
  severity: AlarmSeverity;
  timestamp_ms: number;
  bt: number;
  et: number;
}

export interface ReplayMessage {
  type: "replay";
  timestamp_ms: number;
  et: number;
  bt: number;
  et_ror: number;
  bt_ror: number;
  controls: Record<string, number>;
  progress_pct: number;
  total_duration_ms: number;
}

export interface ControlAckMessage {
  type: "control_ack";
  channel: string;
  value: number;
  applied: boolean;
  enabled: boolean;
  message: string;
}

export interface ErrorMessage {
  type: "error";
  code: string;
  message: string;
  recoverable: boolean;
}

export interface ConnectionMessage {
  type: "connection";
  driver_state: DriverState;
  driver_name: string;
  message: string;
}

export type ServerMessage =
  | TemperatureMessage
  | EventMessage
  | StateMessage
  | AlarmMessage
  | ReplayMessage
  | ControlAckMessage
  | ErrorMessage
  | ConnectionMessage;

// ──────────────────────────────────────────────
// Client → Server messages
// ──────────────────────────────────────────────

export interface ControlCommand {
  type: "control";
  channel: string;
  value: number; // 0.0 – 1.0
  enabled?: boolean;
}

export type CommandAction =
  | "start_monitoring"
  | "stop_monitoring"
  | "start_recording"
  | "stop_recording"
  | "mark_event"
  | "reset"
  | "sync";

export interface SessionCommand {
  type: "command";
  action: CommandAction;
  event_type?: RoastEventType; // required when action === 'mark_event'
  last_timestamp_ms?: number; // required when action === 'sync'
}

export type ReplayActionType = "start" | "pause" | "resume" | "stop" | "seek";

export interface ReplayControlCommand {
  type: "replay_control";
  action: ReplayActionType;
  profile_id?: string;
  speed?: number; // 0 < speed <= 10, default 1.0
  timestamp_ms?: number; // required when action === 'seek'
}

export type ClientMessage =
  | ControlCommand
  | SessionCommand
  | ReplayControlCommand;

File: backend/src/openroast/models/ws_messages.py
"""Pydantic models for WebSocket message serialization.

These models define the wire contract between backend and frontend.
Both sides MUST stay in sync — the TypeScript counterpart is at
frontend/src/lib/types/ws-messages.ts.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field

# ──────────────────────────────────────────────
# Shared enums
# ──────────────────────────────────────────────

class RoastEventType(StrEnum):
    """Known roast event types."""

    CHARGE = "CHARGE"
    DRY_END = "DRY_END"
    FCs = "FCs"
    FCe = "FCe"
    SCs = "SCs"
    SCe = "SCe"
    DROP = "DROP"
    COOL = "COOL"
    TP = "TP"


class SessionStateValue(StrEnum):
    """Session state values matching core.session.SessionState."""

    IDLE = "idle"
    MONITORING = "monitoring"
    RECORDING = "recording"
    FINISHED = "finished"


class AlarmSeverity(StrEnum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class DriverStateValue(StrEnum):
    """Driver connection states matching drivers.base.ConnectionState."""

    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    CONNECTED = "connected"
    ERROR = "error"


class CommandAction(StrEnum):
    START_MONITORING = "start_monitoring"
    STOP_MONITORING = "stop_monitoring"
    START_RECORDING = "start_recording"
    STOP_RECORDING = "stop_recording"
    MARK_EVENT = "mark_event"
    RESET = "reset"
    SYNC = "sync"


class ReplayAction(StrEnum):
    START = "start"
    PAUSE = "pause"
    RESUME = "resume"
    STOP = "stop"
    SEEK = "seek"


# ──────────────────────────────────────────────
# Server → Client messages
# ──────────────────────────────────────────────

class TemperatureMessage(BaseModel):
    """Periodic temperature reading pushed at sampling interval."""

    type: Literal["temperature"] = "temperature"
    timestamp_ms: float = Field(ge=0)
    et: float
    bt: float
    et_ror: float = Field(default=0.0, description="ET rate of rise (C/min)")
    bt_ror: float = Field(default=0.0, description="BT rate of rise (C/min)")
    extra_channels: dict[str, float] = Field(default_factory=dict)


class EventMessage(BaseModel):
    """Roast event notification."""

    type: Literal["event"] = "event"
    event_type: RoastEventType
    timestamp_ms: float = Field(ge=0)
    auto_detected: bool = False
    bt_at_event: float = 0.0
    et_at_event: float = 0.0


class StateMessage(BaseModel):
    """Session state transition notification."""

    type: Literal["state"] = "state"
    state: SessionStateValue
    previous_state: SessionStateValue


class AlarmMessage(BaseModel):
    """Alarm trigger notification."""

    type: Literal["alarm"] = "alarm"
    alarm_id: str
    message: str
    severity: AlarmSeverity
    timestamp_ms: float = Field(ge=0)
    bt: float = 0.0
    et: float = 0.0


class ReplayMessage(BaseModel):
    """Profile replay data point."""

    type: Literal["replay"] = "replay"
    timestamp_ms: float = Field(ge=0)
    et: float
    bt: float
    et_ror: float = 0.0
    bt_ror: float = 0.0
    controls: dict[str, float] = Field(default_factory=dict)
    progress_pct: float = Field(ge=0, le=100)
    total_duration_ms: float = Field(ge=0)


class ControlAckMessage(BaseModel):
    """Acknowledgement of a control command."""

    type: Literal["control_ack"] = "control_ack"
    channel: str
    value: float
    applied: bool
    enabled: bool = True
    message: str = ""


class ErrorMessage(BaseModel):
    """Error notification (non-fatal)."""

    type: Literal["error"] = "error"
    code: str
    message: str
    recoverable: bool = True


class ConnectionMessage(BaseModel):
    """Driver connection state change."""

    type: Literal["connection"] = "connection"
    driver_state: DriverStateValue
    driver_name: str = ""
    message: str = ""


# Union type for dispatch
ServerMessage = (
    TemperatureMessage
    | EventMessage
    | StateMessage
    | AlarmMessage
    | ReplayMessage
    | ControlAckMessage
    | ErrorMessage
    | ConnectionMessage
)


# ──────────────────────────────────────────────
# Client → Server messages
# ──────────────────────────────────────────────

class ControlCommand(BaseModel):
    """Set a control slider value."""

    type: Literal["control"] = "control"
    channel: str
    value: float = Field(ge=0.0, le=1.0)
    enabled: bool = True


class SessionCommand(BaseModel):
    """Session lifecycle or event marking command."""

    type: Literal["command"] = "command"
    action: CommandAction
    event_type: RoastEventType | None = None  # required when action == mark_event
    last_timestamp_ms: float | None = None  # required when action == sync


class ReplayControlCommand(BaseModel):
    """Control profile replay playback."""

    type: Literal["replay_control"] = "replay_control"
    action: ReplayAction
    profile_id: str = ""
    speed: float = Field(default=1.0, gt=0, le=10.0)
    timestamp_ms: float | None = None  # required when action == seek


# Union type for dispatch
ClientMessage = ControlCommand | SessionCommand | ReplayControlCommand
